x-pack/filebeat/otel/exporter_factory.go

```diff
@@ -0,0 +1,208 @@
+		ConsoleOptions: []stdoutmetric.Option{
+			stdoutmetric.WithPrettyPrint(),
+			stdoutmetric.WithTemporalitySelector(DeltaSelector),
+			stdoutmetric.WithEncoder(&ConcurrentEncoder{Encoder: json.NewEncoder(os.Stdout)}),
+		},
+		HttpOptions: []otlpmetrichttp.Option{otlpmetrichttp.WithTemporalitySelector(DeltaSelector)},
+	}
+}
+
+// The MetricsExporterFactory reads environment variables and creates the metrics exporter requested
+// by the environment.
+//
+// To export OTEL metrics to an OTLP/gRPC endpoint set these environment variables:
+//   - OTEL_EXPORTER_OTLP_ENDPOINT: Required. The OTLP endpoint URL.
+//   - OTEL_EXPORTER_OTLP_HEADERS: Required if endpoint is authenticated.
+//
+// To export using the http/protobuf protocol add this environment variable with the above environment variables:
+//   - OTEL_EXPORTER_OTLP_METRICS_PROTOCOL="http/protobuf"
+//
+// To override other environment variables and print a JSON representation of the metrics to console set:
+//   - OTEL_METRICS_EXPORTER=console
+//
+// To override other environment variables and disable OTEL metrics collection set:
+//   - OTEL_METRICS_EXPORTER=none
+type MetricsExporterFactory struct {
+	lock                  sync.Mutex
+	globalMetricsExporter sdkmetric.Exporter
+	exporterOptions       MetricExporterOptions
+}
+
+// initializeGlobalMetricsExporterFactory creates the global exporter factory
+func initializeGlobalMetricsExporterFactory() {
+	exporterFactory = NewMetricsExporterFactory(GetDefaultMetricExporterOptions())
+}
+
+// GetGlobalMetricsExporterFactory returns a globally defined MetricsExporterFactory.
+// The GlobalExporterFactory returns the same Exporter for every call to
+// GetExporter(ctx, true). The global MetricsExporterFactory is created using
+// default MetricExporterOptions.
+// Using the same Exporter across inputs reduces connections to the OTLP endpoint.
+func GetGlobalMetricsExporterFactory() *MetricsExporterFactory {
+	exporterOnce.Do(initializeGlobalMetricsExporterFactory) // Guarantees initializeConfig runs exactly once
+	return exporterFactory
+}
+
+// NewMetricsExporterFactory creates a new MetricsExporterFactory
+// exporterOptions MetricExporterOptions: the options to use for creating the Exporter
+func NewMetricsExporterFactory(exporterOptions MetricExporterOptions) *MetricsExporterFactory {
+	return &MetricsExporterFactory{
+		globalMetricsExporter: nil,
+		exporterOptions:       exporterOptions,
+	}
+}
+
+// GetExporter returns a metrics exporter based on the current environment
+// configuration.
+//
+// If global is true, GetExporter returns the cached global exporter, creating it
+// on first use. If global is false, GetExporter always creates and returns a new
+// exporter.
+func (ef *MetricsExporterFactory) GetExporter(ctx context.Context, global bool) (sdkmetric.Exporter, ExporterType, error) {
+	exporterType := GetExporterTypeFromEnv()
+	var err error
+	if global && ef.globalMetricsExporter != nil {
```

Comment:
`GetExporter()` is still racey, it reads `ef.globalMetricsExporter` outside the mutex, and writes it under the mutex. Please move the “return cached exporter” check inside the locked region